tools/tfdoc.py

```diff
@@ -14,274 +14,350 @@
-  yield '|---|---|:---:|'
-  for o in outputs:
-    yield '| {name} | {description} | {sensitive} |'.format(
-        name=o.name, description=o.description,
-        sensitive='✓' if o.sensitive else '')
-
-
-def format_type(type_spec):
-  "Format variable type."
-  if not type_spec:
-    return ''
+def format_doc(outputs, variables, files, show_extra=True):
+  'Return formatted document.'
   buffer = []
-  stack = []
-  for t in RE_TYPE.split(type_spec.split("\n")[0]):
-    if not t:
-      continue
-    if t in '({':
-      stack.append(t)
-    elif t in '})':
-      stack.pop()
-    buffer.append(t)
-  for t in reversed(stack):
-    buffer.append(')' if t == '(' else '}')
-  return ''.join(buffer).replace('object({})', 'object({...})')
-
-
-def format_variables(variables, required_first=True):
-  "Format variables."
-  if not variables:
+  if files:
+    buffer.append('\n## Files\n')
+    for line in format_files(files):
+      buffer.append(line)
+    buffer.append('\n')
+  if variables:
+    buffer.append('\n## Variables\n')
+    for line in format_variables(variables, show_extra):
+      buffer.append(line)
+  if outputs:
+    buffer.append('\n## Outputs\n')
+    for line in format_outputs(outputs, show_extra):
+      buffer.append(line)
+  buffer.append('\n')
+  return '\n'.join(buffer)
+
+
+def format_files(items):
+  'Format files table.'
+  items.sort(key=lambda i: i.name)
+  yield '| name | description | modules | resources |'
+  yield '|---|---|---|---|'
+  for i in items:
+    modules = resources = ''
+    if i.modules:
+      modules = '<code>%s</code>' % '</code> · <code>'.join(
+          sorted(i.modules))
+    if i.resources:
+      resources = '<code>%s</code>' % '</code> · <code>'.join(
+          sorted(i.resources))
+    yield f'| [{i.name}](./{i.name}) | {i.description} | {modules} | {resources} |'
+
+
+def format_outputs(items, show_extra=True):
+  'Format outputs table.'
+  if not items:
+    return
+  items.sort(key=lambda i: i.name)
+  yield '| name | description | sensitive |' + (
+      ' consumers |' if show_extra else ''
+  )
+  yield '|---|---|:---:|' + (
+      '---|' if show_extra else ''
+  )
+  for i in items:
+    consumers = i.consumers or ''
+    if consumers:
+      consumers = '<code>%s</code>' % '</code> · <code>'.join(
+          consumers.split())
+    sensitive = '✓' if i.sensitive else ''
+    format = f'| {i.name} | {i.description or ""} | {sensitive} |'
+    format += f' {consumers} |' if show_extra else ''
+    yield format
+
+
+def format_variables(items, show_extra=True):
+  'Format variables table.'
+  if not items:
     return
-  variables.sort(key=lambda v: v.name)
-  variables.sort(key=lambda v: v.required, reverse=True)
-  yield '| name | description | type | required | default |'
-  yield '|---|---|:---: |:---:|:---:|'
-  row = (
-      '| {name} | {description} | <code title="{type_spec}">{type}</code> '
-      '| {required} | {default} |'
+  items.sort(key=lambda i: i.name)
```

Comment:
probably not a big deal but you're modifying items. Same thing in `format_outputs`

Reply:
good catch, but I'd say we don't much care as everything is self-contained and items are single use